pkg/descheduler/strategies/duplicates.go

```diff
@@ -202,6 +210,67 @@ func RemoveDuplicatePods(
 	}
 }
 
+func getNodeAffinityNodeSelector(pod *v1.Pod) *v1.NodeSelector {
+	if pod.Spec.Affinity == nil {
+		return nil
+	}
+	if pod.Spec.Affinity.NodeAffinity == nil {
+		return nil
+	}
+	return pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution
+}
+
+func getTargetNodes(podNodes map[string][]*v1.Pod, nodes []*v1.Node) []*v1.Node {
+	// For each duplicated pod get a list of nodes where it can land
+	// distinct pods which differ in taints, node affinity and node selector
+	distinctPods := map[*v1.Pod]struct{}{}
+	for _, pods := range podNodes {
+		for _, pod := range pods {
+			duplicated := false
+			for dp := range distinctPods {
+				if utils.TolerationsEqual(pod.Spec.Tolerations, dp.Spec.Tolerations) &&
+					utils.NodeSelectorsEqual(getNodeAffinityNodeSelector(pod), getNodeAffinityNodeSelector(dp)) &&
+					reflect.DeepEqual(pod.Spec.NodeSelector, dp.Spec.NodeSelector) {
+					duplicated = true
+					continue
```

Comment:
will this count pods as duplicate if they don't have tolerations/selectors/affinity?

Reply:
also could you explain a bit more about what this check is for? I am a little confused about the check for duplicate node selector rules here (not that it's wrong, it's likely just me 🙂 )

Reply:
The idea is to assign each pod a (tolerations, nodeselectors, node affinity) triple (including nil instances). Also, there's some logic that reduces duplicated tolerations/nodeselectors/nodeaffinities terms and sorts the terms so it's easy to identify the duplicated terms. So when two triples are equal, evaluation of both corresponding pods wrt. feasible nodes will end up the same. Thus, reducing the number of (pod, node) evaluations with the same result. Thus, minimizing the set of pods where we check how many nodes are feasible for them once evicted. `getTargetNodes` is computing how many nodes are feasible for all the pods there were marked for eviction. So the average number of pods per node can be adjusted based what `getTargetNodes` returns.

The checks in `getTargetNodes` will be most probably all duplicates of checks in `IsEvictable`.

Reply:
Ah, I see so this check is meant for efficiency in the later checks if I understand correctly?

Reply:
Right, efficiency purposes only.

Reply:
I think that's a great idea. Would you mind adding that to the comment for this block, to explain that this is for efficiency? It wasn't clear to me why we were looking for distinct pods at this point

Reply:
Done